File: leetcode/172_factorial_trailing_zeroes.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function


class Solution(object):
    def trailingZeroes(self, n):
        result = 0

        i, powers = 1, []
        while 5 ** i <= n:
            powers.append(5 ** i)
            i += 1

        for power in powers:
            result += int(n / power)

        return result


if __name__ == '__main__':
    solution = Solution()

    assert 0 == solution.trailingZeroes(3)
    assert 1 == solution.trailingZeroes(5)
    assert 24 == solution.trailingZeroes(100)

File: leetcode/172_factorial_trailing_zeroes.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>

using namespace std;

class Solution {
 public:
    int trailingZeroes(int n) {
        int result = 0;

        while (n > 0) {
            result += n / 5;
            n /= 5;
        }

        return result;
    }
};

int main() {
    auto solution = Solution();

    assert(0 == solution.trailingZeroes(3));
    assert(1 == solution.trailingZeroes(5));
    assert(24 == solution.trailingZeroes(100));
    assert(452137076 == solution.trailingZeroes(1808548329));
}
